src/mds/MDSPerfMetricTypes.h

```diff
@@ -0,0 +1,189 @@
+    return os;
+  }
+};
+
+struct ReadLatencyMetric {
+  utime_t lat;
+
+  DENC(ReadLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const ReadLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct WriteLatencyMetric {
+  utime_t lat;
+
+  DENC(WriteLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const WriteLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct MetadataLatencyMetric {
+  utime_t lat;
+
+  DENC(MetadataLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const MetadataLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+WRITE_CLASS_DENC(CapHitMetric)
+WRITE_CLASS_DENC(ReadLatencyMetric)
+WRITE_CLASS_DENC(WriteLatencyMetric)
+WRITE_CLASS_DENC(MetadataLatencyMetric)
+
+// metrics that are forwarded to the MDS by client(s).
+struct Metrics {
+  // metrics
+  CapHitMetric cap_hit_metric;
+  ReadLatencyMetric read_latency_metric;
+  WriteLatencyMetric write_latency_metric;
+  MetadataLatencyMetric metadata_latency_metric;
+
+  // metric update type
+  UpdateType update_type = UpdateType::UPDATE_TYPE_REFRESH;
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+
+    ENCODE_START(1, 1, bl);
+    encode(static_cast<uint32_t>(update_type), bl);
+    encode(cap_hit_metric, bl);
+    encode(read_latency_metric, bl);
+    encode(write_latency_metric, bl);
+    encode(metadata_latency_metric, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+
+    DECODE_START(1, iter);
+    uint32_t type;
+    decode(type, iter);
+    update_type = static_cast<UpdateType>(type);
+    decode(cap_hit_metric, iter);
+    decode(read_latency_metric, iter);
+    decode(write_latency_metric, iter);
+    decode(metadata_latency_metric, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_int("update_type", static_cast<uint32_t>(update_type));
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const Metrics& metrics) {
+    os << "[update_type=" << metrics.update_type << ", metrics={"
+       << "cap_hit_metric=" << metrics.cap_hit_metric
+       << ", read_latency=" << metrics.read_latency_metric
+       << ", write_latency=" << metrics.write_latency_metric
+       << ", metadata_latency=" << metrics.metadata_latency_metric
+       << "}]";
+    return os;
+  }
+};
+
+// Initial Sequence Number (ISN) of the first ping message sent
+// by rank 0 to other active ranks (incuding itself).
+#define MDS_PINGER_ISN 1
+
+struct MetricsMessage {
```

Comment:
if it's a plain struct for data, might want to name it like `metrics_message_t`.

Reply:
+1 MetricsMessage makes you think it comes from `src/messages`.

Reply:
ACK